metadata/src/lib.rs

```diff
@@ -641,35 +642,62 @@ pub struct RuntimeApiMethodParamMetadata {
 /// Metadata of custom types with custom values, basically the same as `frame_metadata::v15::CustomMetadata<PortableForm>>`.
 #[derive(Debug, Clone)]
 pub struct CustomMetadata {
+    /// custom values accessible via a name.
     map: BTreeMap<String, frame_metadata::v15::CustomValueMetadata<PortableForm>>,
 }
 
 impl CustomMetadata {
-    /// Get a certain [CustomMetadataValue] by its name.
-    pub fn get(&self, name: &str) -> Option<CustomMetadataValue<'_>> {
-        self.map.get(name).map(|e| CustomMetadataValue {
+    /// Get a certain [CustomValueMetadata] by its name.
+    pub fn get<'a>(&'a self, name: &'a str) -> Option<CustomValueMetadata<'_>> {
+        self.map.get(name).map(|e| CustomValueMetadata {
             type_id: e.ty.id,
             data: &e.value,
+            name,
+        })
+    }
+
+    /// Iterates over names (keys) and associated custom values
+    pub fn iter(&self) -> impl Iterator<Item = CustomValueMetadata> {
+        self.map.iter().map(|(name, el)| CustomValueMetadata {
+            type_id: el.ty.id,
+            data: &el.value,
+            name: name.as_ref(),
         })
     }
 }
 
 /// Basically the same as `frame_metadata::v15::CustomValueMetadata<PortableForm>>`, but borrowed.
-pub struct CustomMetadataValue<'a> {
+pub struct CustomValueMetadata<'a> {
     type_id: u32,
     data: &'a [u8],
+    name: &'a str,
 }
 
-impl<'a> CustomMetadataValue<'a> {
-    /// the scale encoded value
+impl<'a> CustomValueMetadata<'a> {
+    /// The scale encoded value
     pub fn bytes(&self) -> &'a [u8] {
         self.data
     }
 
-    /// the type id in the TypeRegistry
+    /// The type id in the TypeRegistry
     pub fn type_id(&self) -> u32 {
         self.type_id
     }
+
+    /// The name under which the custom value is registered.
+    pub fn name(&self) -> &str {
+        self.name
+    }
+
+    /// Calculates the hash for the CustomValueMetadata.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `self.type_id` is not registered in the provided type registry
+    pub fn get_hash(&self, registry: &PortableRegistry) -> [u8; HASH_LEN] {
```

Comment:
Ah, and, have a look at how we get the hashes for eg runtime APIs above. I think we should:

- Rename this to `hash` for consistency
- Keep a copy of the portable registry internally, so the user doesn't need to provide one (which doesn't make much sense since this is the metadata which should have all of that, and opens the door for inconsistency :))